helix-core/src/main/java/org/apache/helix/controller/rebalancer/util/DelayedRebalanceUtil.java

```diff
@@ -0,0 +1,215 @@
+package org.apache.helix.controller.rebalancer.util;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import org.apache.helix.model.ClusterConfig;
+import org.apache.helix.model.IdealState;
+import org.apache.helix.model.InstanceConfig;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * The util to support delayed rebalance logic.
+ */
+public class DelayedRebalanceUtil {
```

Comment:
General comment on this util class:

Tip: Instead of passing in null as a parameter, just create another method (overloaded) without that parameter. That way you get to avoid passing nulls and the caller doesn't have to pass nulls around.

Reply:
This is mainly for avoiding duplicate code. I thought about this, I think we can do it reverse way. The method takes both instance config and cluster config shall call the method takes cluster config only. Then do an additional check based on that. I will change it.

Reply:
Either way there's some duplication. If you absolutely have to have duplication, we'd like to make it as clean as possible :)